multi_gpu/four/model_parallel.py

```diff
@@ -0,0 +1,162 @@
+####################
+"""Model Adapted from https://github.com/jamespengcheng/PyTorch-CNN-on-CIFAR10/blob/master/ConvNetClassifier.py"""
+####################
+
+
+import torch
+import torchvision
+import torchvision.datasets as datasets
+import torchvision.transforms as transforms
+from torch.autograd import Variable
+import torch.nn as nn
+import torch.nn.functional as F
+
+#import wandb
+
+#wandb.init(project="FOUR")
+
+# Configuration area
+#############################
+epochs = 10
+batch_size = 256
+learning_rate = 0.001
+num_workers = 4
+#############################
+use_cuda = True
+device = torch.cuda.device("cuda" if use_cuda else "cpu")
+
+transform_train = transforms.Compose([
+    transforms.RandomCrop(32, padding=4),
+    transforms.RandomHorizontalFlip(),
+    transforms.ToTensor(),
+    transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
+])
+
+transform_test = transforms.Compose([
+    transforms.ToTensor(),
+    transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010)),
+])
+
+cifar_trainset = datasets.CIFAR10(root='./data', train=True, download=True, transform=transform_train)
+trainset_loader = torch.utils.data.DataLoader(cifar_trainset, batch_size=batch_size,shuffle=True, num_workers=num_workers)
+
+cifar_testset = datasets.CIFAR10(root='./data', train=False, download=True, transform=transform_test)
+testset_loader = torch.utils.data.DataLoader(cifar_testset, batch_size=batch_size, shuffle=False, num_workers=num_workers)
+
+class ConvNet(nn.Module):
+    def __init__(self):
+        super(ConvNet, self).__init__()
+        self.conv1 = nn.Conv2d(in_channels=3, out_channels=48, kernel_size=(3,3), padding=(1,1))
+        self.conv1 = nn.DataParallel(self.conv1)
```

Comment:
DataParallel is easy to use, but is inferior to DistributedDataParallel which uses multiprocessing instead of multithreading. This tutorial provides a nice overview that you can apply to this code: https://medium.com/analytics-vidhya/distributed-training-in-pytorch-part-1-distributed-data-parallel-ae5c645e74cb
You probably only need to have these examples function for 1 node with multiple GPUs which should simplify things. The 'MASTER_ADDRESS' and 'MASTER_PORT' environmental variables might not matter if you are only using one node, but you can play around with that.